text/0000-delegation.md

```diff
@@ -0,0 +1,700 @@
+
+```rust
+impl TR for S {
+    type Item = <F as TR>::Item;
+    // ...
+    const MAX = <F as TR>::MAX;
+    // ...
+    fn foo(&self) -> u32 {
+        self.f.foo()
+    }
+    fn bar(&self, x: u32, y: u32, z: u32) -> u32 {
+        self.f.bar(x, y, z)
+    }
+    // ...
+}
+```
+
+To delegate most of a trait, rather than all of it, simply `delegate *` and then write the manual implementations for the items you don’t want to delegate.
+
+```rust
+impl TR for S {
+    delegate * to f;
+
+    fn foo(&self) -> u32 {
+        42
+    }
+}
+```
+
+Aside from the implementation of `foo()`, this has exactly the same meaning as the first example.
+
+If you only want to delegate specific items, rather than “all” or “most” items, then replace `*` with a comma-separated list of only the items you want to delegate. Since it’s possible for types and functions to have the same name, the items must be prefixed with `fn`, `const` and `type` as appropriate.
+
+```rust
+impl TR for S {
+    delegate fn foo, fn bar, const MAX, type Item
+        to f;
+}
+```
+
+This also has the exact same meaning as the first example. While `to field_name;` is not required to be on a new line, this is the `rustfmt` default when delegating individual trait items, as it makes it easy to visually identify the field being delegated to.
+
+If you want to delegate some methods to one field and some to another, simply write multiple `delegate` statements:
+
+```rust
+impl TR for S {
+    delegate * to field_one;
+    delegate fn foo, const MAX, type Item
+        to field_two;
+}
+```
+
+This delegates `foo`, `MAX`, and `Item` to `field_two`. Everything else is delegated to `field_one`.
+
+In addition to "regular" structs with named fields, you can delegate to indexed fields on a tuple struct. This first field name is `0`, which we're used to writing as `self.0`:
+
+```rust
+struct AddOnlyCounter(u32);
+
+impl AddOnlyCounter {
+    fn new() -> Self {
+        Self { 0 }
+    }
+}
+
+impl AddAssign for AddOnlyCounter {
+    delegate * to 0;
+}
+```
+
+By using the _newtype pattern_, we _restrict_ the behaviour of the original type and only allow semantically valid methods/operations. If `AddOnlyCounter` is defined in another module, users can't break its guarantees, as the `u32` is private aka encapsulated.
+
+```rust
+let counter = AddOnlyCounter::new();
+
+// this works:
+counter += 1;
+
+// these are compile-time errors:
+counter = 5;
+counter -= 1;
+counter.0 -= 1;
+```
+
+
+# Reference-level explanation
+[reference]: #reference
+
+A delegation statement can only appear inside a trait impl block. Delegation inside inherent impls is left as a future extension.
+
+Delegation must be to a field on `Self`. Other kinds of implementer expressions are left as future extensions. This also means delegation can only be done on structs for now.
+
+There may be more than one delegation statement. For readability, `rustfmt` moves delegation statements to the top of an impl block.
```

Comment:
Perhaps justify _why_ this is more readable?  I can see an argument for `*`, but especially for single-method delegation, I'm not convinced.  Particularly in a case where I'm changing existing code to replace manual delegation with `delegate`, I'd expect rustfmt to leave the the item where it is so the diff is the obvious one.  And even in new code, I might be intentionally matching the order of the methods on the trait, for example.  (And more nit-picky, if it's going to move multiple, it would need to pick an order to put them in.)

Also, what's the semantics (since this is the reference section) of two `delegate * to` items?

Reply:
My thinking here is that:
+ The argument starts from `*`, which should be at the top so that it is seen first
+ For consistency with `*`, you place all delegation where `delegate * to whatever` would be

An argument could however be made that all delegations should be at the bottom since they often will be `delegate *` and so they are a sort of "catch the rest", i.e, they function like `match x { important => .. , _ => .. }` does wrt. `_ =>`.

With respect to order, I'd first group items by item type and then in each group alphabetically so:
+ const
+ type
+ fn

Reply:
Interesting!

PS: we could leave formatting bikeshed up to a style-fmt RFC.

Reply:
:+1: that's why https://github.com/rust-lang-nursery/fmt-rfcs exists